ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/PointSchedule/Shrinking/Tests.hs

```diff
@@ -0,0 +1,104 @@
+
+import           Data.Foldable (toList)
+import           Data.Map (keys)
+import           Data.Maybe (mapMaybe)
+import           Test.Consensus.Genesis.Setup (genChains)
+import           Test.Consensus.Genesis.Tests.Uniform (genUniformSchedulePoints)
+import           Test.Consensus.PointSchedule (PeerSchedule, PeersSchedule,
+                     prettyPeersSchedule)
+import           Test.Consensus.PointSchedule.Peers (Peer (..), Peers (..))
+import           Test.Consensus.PointSchedule.Shrinking (shrinkHonestPeer)
+import           Test.Consensus.PointSchedule.SinglePeer (SchedulePoint (..))
+import           Test.QuickCheck (Property, conjoin, counterexample)
+import           Test.Tasty
+import           Test.Tasty.QuickCheck (choose, forAllBlind, testProperty)
+import           Test.Util.TestBlock (TestBlock)
+
+tests :: TestTree
+tests =
+  testGroup "shrinking functions"
+    [ testGroup "honest peer shrinking"
+      [ testProperty "actually shortens the schedule" prop_shortens
+      , testProperty "preserves the final state all peers" prop_preservesFinalStates
+      , testProperty "doesn't remove points of the adversarial schedule" prop_preserversAdversarial
+      ]
+    ]
+
+prop_shortens :: Property
+prop_shortens = checkShrinkProperty isShorterThan
+
+prop_preservesFinalStates :: Property
+prop_preservesFinalStates = checkShrinkProperty doesNotChangeFinalState
+
+prop_preserversAdversarial :: Property
+prop_preserversAdversarial = checkShrinkProperty doesNotRemoveAdversarialPoints
+
+-- | Apparently, `unsnoc` hasn't been invented yet, so we'll do this manually
+lastM :: [a] -> Maybe a
+lastM []     = Nothing
+lastM [a]    = Just a
+lastM (_:ps) = lastM ps
+
+samePeers :: PeersSchedule blk -> PeersSchedule blk -> Bool
+samePeers sch1 sch2 = (keys $ others sch1) == (keys $ others sch2)
+
+isShorterThan :: PeersSchedule blk -> PeersSchedule blk -> Bool
+isShorterThan original shrunk =
+  samePeers original shrunk
+  && (or $ zipWith
+    (\oldSch newSch -> (fst <$> lastM newSch) < (fst <$> lastM oldSch))
+    (toList original)
+    (toList shrunk)
+  )
+
+doesNotChangeFinalState :: Eq blk => PeersSchedule blk -> PeersSchedule blk -> Bool
+doesNotChangeFinalState original shrunk =
+  samePeers original shrunk
+  && (and $ zipWith
+    (\oldSch newSch ->
+      lastTP oldSch == lastTP newSch &&
+      lastHP oldSch == lastHP newSch &&
+      lastBP oldSch == lastBP newSch
+    )
+    (toList original)
+    (toList shrunk)
+  )
+  where
+    lastTP :: PeerSchedule blk -> Maybe (SchedulePoint blk)
+    lastTP sch = lastM $ mapMaybe (\case (_, p@(ScheduleTipPoint    _)) -> Just p ; _ -> Nothing) sch
+    lastHP :: PeerSchedule blk -> Maybe (SchedulePoint blk)
+    lastHP sch = lastM $ mapMaybe (\case (_, p@(ScheduleHeaderPoint _)) -> Just p ; _ -> Nothing) sch
+    lastBP :: PeerSchedule blk -> Maybe (SchedulePoint blk)
+    lastBP sch = lastM $ mapMaybe (\case (_, p@(ScheduleBlockPoint  _)) -> Just p ; _ -> Nothing) sch
```

Comment:
Those helpers used to exist at some point in the `PointSchedule` module, but maybe not anymore?